### PR TITLE
IGNITE-20826 Add maxObservableSafeTime initialization on leader election

### PR DESCRIPTION
**Pull request by sanpwc**
https://issues.apache.org/jira/browse/IGNITE-20826

maxObservableSafeTime is now evaluated as clock.now() + CLOCK_SKEW on each leader re-election, thus it's no longer possible to have missvalidated safeTimePropagation command that was skipped because of uninitialized watermark.

Given PR also contain several unrelates mixins:
1. Storage scan on partition restart is removed. It's no longer needed because of fully implement write intent resolution logic, including the one on recovery.
2. LocalSnapshotCopierTest#testInterrupt() enabled.
3. Obsolete closed todo removed from the code. // TODO: need to properly handle primary replica changes https://issues.apache.org/jira/browse/IGNITE-17615

### CODE REVIEW

File: modules/table/src/integrationTest/java/org/apache/ignite/distributed/ReplicasSafeTimePropagationTest.java [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+        // Stop the leader - the only node that actually do safeTime watermark validation within onBeforeApply.
+        assertNotNull(raftClient.leader());
+
+        PartialNode nodeTopStop = cluster.get(raftClient.leader().consistentId());
+
+        assertNotNull(nodeTopStop);
+
+        nodeTopStop.stop();
+
+        // Select alive raft client
```
Comment (ibessonov) [inline modules/table/src/integrationTest/java/org/apache/ignite/distributed/ReplicasSafeTimePropagationTest.java:145]:
```suggestion
        // Select alive raft client.
```

File: modules/table/src/integrationTest/java/org/apache/ignite/distributed/ReplicasSafeTimePropagationTest.java [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+        RaftGroupService raftClient = someNode.raftClient;
+
+        assertThat(raftClient.refreshLeader(), willCompleteSuccessfully());
+
+        long firstSafeTime = calculateSafeTime(someNode.clock);
+
+        // Send command with safe time X.
+        sendSafeTimeSyncCommand(raftClient, firstSafeTime, false);
+
+        // Stop all nodes
```
Comment (ibessonov) [inline modules/table/src/integrationTest/java/org/apache/ignite/distributed/ReplicasSafeTimePropagationTest.java:188]:
```suggestion
        // Stop all nodes.
```

File: modules/table/src/integrationTest/java/org/apache/ignite/distributed/ReplicasSafeTimePropagationTest.java [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+    ) {
+        CompletableFuture<Object> safeTimeCommandFuture = raftClient.run(
+                REPLICA_MESSAGES_FACTORY
+                        .safeTimeSyncCommand()
+                        .safeTimeLong(safeTime)
+                        .build()
+        );
+
+        if (expectSafeTimeReorderException) {
+            assertThat(safeTimeCommandFuture, CompletableFutureExceptionMatcher.willThrow(SafeTimeReorderException.class));
```
Comment (ibessonov) [inline modules/table/src/integrationTest/java/org/apache/ignite/distributed/ReplicasSafeTimePropagationTest.java:228]: Please import this static method :)

File: modules/table/src/integrationTest/java/org/apache/ignite/distributed/ReplicasSafeTimePropagationTest.java [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+     *     <li>Restart the cluster.</li>
+     *     <li>Send command with safe time less than previously applied to the leader before the restart
+     *     and verify that SafeTimeReorderException is thrown.</li>
+     * </ol>
+     */
+    @Test
+    public void testSafeTimeReorderingOnLeaderRestart() throws Exception {
+        // Start two node and a raft group with two peer.
+        {
+            cluster = Set.of("node1", "node2").parallelStream().collect(Collectors.toMap(Function.identity(), PartialNode::new));
```
Comment (ibessonov) [inline modules/table/src/integrationTest/java/org/apache/ignite/distributed/ReplicasSafeTimePropagationTest.java:172]: I don't think that invoking a constructor required different thread, what was your motivation?

File: modules/table/src/integrationTest/java/org/apache/ignite/distributed/ReplicasSafeTimePropagationTest.java [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+
+        private RaftGroupService raftClient;
+
+        PartialNode(String nodeName) {
+            this.nodeName = nodeName;
+            this.clock = new HybridClockImpl();
+        }
+
+        CompletableFuture<Void> start() throws Exception {
+            clusterService = ClusterServiceTestUtils.clusterService(nodeName, port++, NODE_FINDER);
```
Comment (ibessonov) [inline modules/table/src/integrationTest/java/org/apache/ignite/distributed/ReplicasSafeTimePropagationTest.java:255]: There's a chance that you wanted to call `start()` in parallel. If that's true then `port++` is unsafe

**Review by ibessonov — APPROVED** [review]
Only minor comments

File: modules/table/src/integrationTest/java/org/apache/ignite/distributed/ReplicasSafeTimePropagationTest.java (hunk 3 above)
Comment (sanpwc) [reply modules/table/src/integrationTest/java/org/apache/ignite/distributed/ReplicasSafeTimePropagationTest.java:228]: Fixed.

File: modules/table/src/integrationTest/java/org/apache/ignite/distributed/ReplicasSafeTimePropagationTest.java (hunk 4 above)
Comment (sanpwc) [reply modules/table/src/integrationTest/java/org/apache/ignite/distributed/ReplicasSafeTimePropagationTest.java:172]: Long story, at the very beginning that was create + start. Fixed.

File: modules/table/src/integrationTest/java/org/apache/ignite/distributed/ReplicasSafeTimePropagationTest.java (hunk 5 above)
Comment (sanpwc) [reply modules/table/src/integrationTest/java/org/apache/ignite/distributed/ReplicasSafeTimePropagationTest.java:255]: Yep, you are right. Reworked to AtomicInteger.

**Review by ibessonov — APPROVED** [review]

**Pull request merged**